File: spyweb/solver/encoding.py
from dataclasses import dataclass

from spyweb.core.model import (
    Answer,
    AnswerCode,
    LandmarkAnswer,
    LandmarkId,
    NothingAnswer,
    Question,
    QuestionId,
    Rules,
    Sense,
    SpyAnswer,
    SpyId,
)


@dataclass(frozen=True)
class Encoding:
    rules: Rules

    @property
    def question_count(self) -> int:
        return len(self.rules.spies) * len(Sense)

    @property
    def nothing(self) -> AnswerCode:
        return AnswerCode(len(self.rules.spies) + len(self.rules.landmarks))

    def question_id(self, question: Question) -> QuestionId:
        return QuestionId(int(question.spy) * len(Sense) + int(question.sense))

    def decode_question(self, question: QuestionId) -> Question:
        return Question(SpyId(int(question) // len(Sense)), Sense(int(question) % len(Sense)))

    def answer_code(self, answer: Answer) -> AnswerCode:
        if isinstance(answer, NothingAnswer):
            return self.nothing
        if isinstance(answer, SpyAnswer):
            return AnswerCode(answer.spy)
        return AnswerCode(len(self.rules.spies) + int(answer.landmark))

    def decode_answer(self, code: AnswerCode) -> Answer:
        if code == self.nothing:
            return NothingAnswer()
        if code < len(self.rules.spies):
            return SpyAnswer(SpyId(code))
        return LandmarkAnswer(LandmarkId(code - len(self.rules.spies)))

File: src/solver/encoding.ts
import {
  answerCode,
  landmarkId,
  questionId,
  senses,
  spyId,
  type Answer,
  type AnswerCode,
  type Question,
  type QuestionId,
  type Rules
} from "../core/model.js";

export interface Encoding {
  readonly nothing: AnswerCode;
  readonly questionCount: number;
  questionId(question: Question): QuestionId;
  decodeQuestion(id: QuestionId): Question;
  answerCode(answer: Answer): AnswerCode;
  decodeAnswer(code: AnswerCode): Answer;
}

export function createEncoding(rules: Rules): Encoding {
  const spyAnswerCount = rules.spies.length;
  const landmarkOffset = spyAnswerCount;
  const nothing = answerCode(spyAnswerCount + rules.landmarks.length);

  return {
    nothing,
    questionCount: rules.spies.length * senses.length,
    questionId: ({ spy, sense }) => questionId(spy * senses.length + senses.indexOf(sense)),
    decodeQuestion: (id) => {
      const spy = Math.floor(id / senses.length);
      const sense = senses[id % senses.length];
      if (sense === undefined) throw new Error(`Invalid question id ${id}`);
      return { spy: spy as Question["spy"], sense };
    },
    answerCode: (answer) => {
      if (answer.kind === "nothing") return nothing;
      return answer.kind === "spy"
        ? answerCode(answer.spy)
        : answerCode(landmarkOffset + answer.landmark);
    },
    decodeAnswer: (code) => {
      if (code === nothing) return { kind: "nothing" };
      if (code < landmarkOffset) return { kind: "spy", spy: spyId(code) };
      return {
        kind: "landmark",
        landmark: landmarkId(code - landmarkOffset)
      };
    }
  };
}
